CultureInfo should independent from local computer settings (Lombiq Technologies: OCORE-86)

hishamco opened the pull request:
Fixes #11228

hishamco commented:
Seems I need to rename `LocalizationOptions` to avoid naming collision

hishamco commented on src/OrchardCore.Modules/OrchardCore.Localization/CultureLocalizationOptions.cs:
```diff
@@ -0,0 +1,6 @@
+namespace OrchardCore.Localization;
+
+public class CultureLocalizationOptions
```
We could rename it to `OrchardCoreLocalizationOptions` naming is hard ;)

hishamco commented on src/OrchardCore.Modules/OrchardCore.Localization/CultureSettings.cs:
```diff
@@ -0,0 +1,7 @@
+namespace OrchardCore.Localization;
+
+public enum CultureSettings
```
Could be moved to `OC.Abstraction` if we want to add it to `ILocalizationService`

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Localization/CultureScope.cs:
```diff
@@ -1,45 +1,55 @@
+
+    private static CultureScope CreateInternal(string culture, string uiCulture)
+    {
+        var useUserSelectedCultureSettings = ShellScope.Services.GetService<ILocalizationService>()
```
@jtkech is there a better way than using `ShellScope.Services`?

jtkech replied:
That's okay for me to resolve a service like this but only when we are sure it is always used in the context of a valid ShellScope.

Then here you have to use .GetAwaiter().GetResult() which is not good, you would need to add an async `CultureScope.CreateAsync()`.

That said here I would have thought about an additional parameter `useUserOverride` that would be passed by the consumer. Also not sure the option should come from admin settings, maybe better to use a configurable IOptions.

hishamco replied:
> but only when we are sure it is always used in the context of a valid ShellScope

How can I verify?

> That said here I would have thought about an additional parameter useUserOverride that would be passed by the consumer. Also not sure the option should come from admin settings, maybe better to use a configurable IOptions.

If you look to the commits history I already used `IOptions<T>` then I move into `ILocalizationService`

Did you think I need to follow the same pattern that I did in `CultureScopeTests` in other broken tests?

@Piedone any though coz `CultureScope` doesn't accept constructor injection

hishamco replied:
> but only when we are sure it is always used in the context of a valid ShellScope

How can I verify?

> That said here I would have thought about an additional parameter useUserOverride that would be passed by the consumer. Also not sure the option should come from admin settings, maybe better to use a configurable IOptions.

If you look to the commits history I already used `IOptions<T>` then I move into `ILocalizationService`

Did you think I need to follow the same pattern that I did in `CultureScopeTests` in other broken tests?

@Piedone any though coz `CultureScope` doesn't accept constructor injection

hishamco replied:
> That said here I would have thought about an additional parameter useUserOverride that would be passed by the consumer

This is could be a simple option, but I was thinking about centeral place that fetch the culture settings either from `appsettings.json` or site settings

Piedone replied:
Yep, I'd use `IOptions` too.

hishamco replied:
If you want to go with `IOptions` as I did at the beginning, I will rollback from the latest changes

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Localization/Models/LocalizationSettings.cs:
```diff
@@ -1,31 +1,36 @@
 using System.Globalization;
 
-namespace OrchardCore.Localization.Models
+namespace OrchardCore.Localization.Models;
```
Please don't update existing files with this because it's breaking the diff in the PR and we can't see where the differences are. Only do it for new files.

hishamco commented:
@sebastienros I just use `CultureSettings` enum to make clear, but I never mind to replace it with a boolean value, one more thing I just moved `CultureLocalizationOptions` into `OC.Localization.Abstractions` all the options that we are using are placed in their modules, but while this kind of option could be use in any localization implementation I think it's fine to place it in `OC.Localization.Abstractions` rather than `OC.Abstractions`. Another thing I suggest to use the system culture instead of user-selected culture by default, similar to what `CultureInfo` is using

If sound looks good I will remove everything I added related to `ILocalizationService`

jtkech commented:
@hishamco

Take care, about `CultureInfo` the user is not an OC user but an user at the machine level, that's why we were talking about System settings.

There are 2 ctors `CultureInfo(string name)` and `CultureInfo(string name, bool useUserOverride)`, the first one is the same as using `CultureInfo(name, useUserOverride: true)`, so if we assume that the default is when using the first ctor, the default is then `useUserOverride = true`.

So the 2 defined values of the current enum `Default, User` is confusing because of the above. Also, I have nothing against using enum but not when it is to retrieve a value to be used as a bool.

So to not be confused with the "user" we could name this option field `UseSystemSettings` whose default would be true, as I remember @sebastienros suggested `IgnoreSystemSettings` but whose default value would be false and we would need to pass `useUserOverride = !option.IgnoreSystemSettings`.

Not sure we need to bind this IOptions to the configuration (e.g. from appsettins.json), if so and if we bind this field individually, if we use `UseSystemSettings` we would need to pass to the `GetValue()` a default value of true, if we bind the whole IOptions class (as we do many time) we just need to specify in the IOptions class a default value of true for this field.

Or if we use `IgnoreSystemSettings`, as said above, we would just need to pass `useUserOverride = !option.IgnoreSystemSettings`

Maybe better to name this IOptions `CultureOptions` and define it in `OC.Abstractions`, where we define `CultureScope`, for example this IOptions would be used each time we use `CultureScope`.

hishamco commented:
> Take care, about CultureInfo the user is not an OC user but an user at the machine level, that's why we were talking about System settings.

Sure

> There are 2 ctors CultureInfo(string name) and CultureInfo(string name, bool useUserOverride), the first one is the same as using CultureInfo(name, useUserOverride: true), so if we assume that the default is when using the first ctor, the default is then useUserOverride = true.

Oops, seem I made a mistake here, I presumed it's a default culture by default

> Or if we use IgnoreSystemSettings, as said above, we would just need to pass useUserOverride = !option.IgnoreSystemSettings

I never mind with this option as I mentioned above

> Maybe better to name this IOptions CultureOptions and define it in OC.Abstractions, where we define CultureScope, for example this IOptions would be used each time we use CultureScope.

It was there before, but while all the modules adding the options in module project and `CultureScope` is using `CultureOptions` that's why I move it into `OC.Localization.Abstractions`

jtkech reviewed (commented):
Mainly params names/logic, otherwise LGTM

hishamco commented on src/OrchardCore.Modules/OrchardCore.Autoroute/Handlers/AutoroutePartHandler.cs:
```diff
@@ -397,7 +397,10 @@ private async Task GenerateContainerPathFromPatternAsync(AutoroutePart part)
+
+                var cultureOptions = _serviceProvider.GetService<IOptions<CultureOptions>>().Value;
+
+                using (CultureScope.Create(cultureAspect.Culture, useUserOverride: !cultureOptions.IgnoreSystemSettings))
```
>  and let the logic be applied at the lowest level in CultureScope and the new extensions helpers.

Do you mean passing the `CultureOption` to the `CultureScope`?

jtkech replied:
If for the options we don't use `UseUserOverride` but `IgnoreSystemSettings`, this to not be confused with the term `User` and to have a default value of `false`, what I meant is to use everywhere params named `IgnoreSystemSettings`.

So here in place of

    using (CultureScope.Create(cultureAspect.Culture, useUserOverride: !cultureOptions.IgnoreSystemSettings))

Would be

    using (CultureScope.Create(cultureAspect.Culture, ignoreSystemSettings: cultureOptions.IgnoreSystemSettings))

So I suggested to apply the logic, here I mean using the `!` operator, only at the lowest level where we use `new CultureInfo()`, here I mean for example in `CultureScope.Create()` and in the new extensions, here I'm reffering to `AddSupportedCultures()` and `AddSupportedUICultures()` that would have the same renamed param `ignoreSystemSettings` in place of `useUserOverride`.

Passing the whole `CultureOptions` is another good idea.

hishamco replied:
I'm still prefering `UseSystemCulture` or `UserSelectedCultureSetttings` rather than `IgnoreSystemSettings`

> Passing the whole `CultureOptions` is another good idea.

I think passing boolean value is much much better, to make the API simple and easy for testing

hishamco commented on src/OrchardCore.Modules/OrchardCore.Localization/Startup.cs:
```diff
@@ -35,6 +43,8 @@ public override void ConfigureServices(IServiceCollection services)
 
             services.Replace(ServiceDescriptor.Singleton<ILocalizationFileLocationProvider, ModularPoFileLocationProvider>());
+
+            services.Configure<CultureOptions>(_shellConfiguration.GetSection("OrchardCore_Localization"));
```
This way the `CultureOptions` will be there even the localization module is disable

hishamco replied:
And that's why it was fine when I use it in `ILocalizationService` before

jtkech replied:
No because the above startup only run if the `OC.Localization` module is enabled and this is the one that also registers `LocalizationService`, in `AddDefaultServices()` we only register `DefaultLocalizationService`.

Not sure what you wanted to do in `ILocalizationService` but would need to be in both `LocalizationService` and `DefaultLocalizationService`. And if you need to inject `IOptions<CultureOptions>` in both of them, then same problem, here `services.Configure<CultureOptions>()` is only registered if `OC.Localization` is enabled.

But maybe I didn't understand what you meant.

jtkech commented on src/OrchardCore/OrchardCore.Abstractions/Localization/CultureScope.cs:
```diff
@@ -3,34 +3,69 @@
-            Culture = culture;
-            UICulture = uiCulture;
+            Culture = new CultureInfo(culture, useUserOverride);
+            UICulture = new CultureInfo(uiCulture, useUserOverride);
```
As said in an above comment

> If for the options we don't use UseUserOverride but IgnoreSystemSettings, this to not be confused with the term User and to have a default value of false, what I meant is to use everywhere params named IgnoreSystemSettings.

So also in the below Create method and then here in this static ctor where you will apply the logic, meaning using the `!` operator, at the lowest level when using `new CultureInfo()` for which we can't change the existing param name.

        private CultureScope(string culture, string uiCulture, bool ignoreSystemSettings)
        {
            Culture = new CultureInfo(culture, useUserOverride: !ignoreSystemSettings)
            UICulture = new CultureInfo(uiCulture, useUserOverride: !ignoreSystemSettings)
            ...

        public static CultureScope Create(CultureInfo culture, bool ignoreSystemSettings = false)
            => Create(culture, culture, ignoreSystemSettings);

jtkech commented:
@hishamco 

If we assume that `IgnoreSystemSettings` is a good name for the configurable option, because it doesn't contain the term `User` which may be confusing as here it is not an OC user, and because it has a default value of false, here I was just suggesting to use this good name as much as possible, also for the names of method params, for example `AnyMethod(... , goodName: options.GoodName)`.

But no problem, was only a simple naming suggestion, otherwise LGTM.

The only thing is that if we provide a concrete value e.g. in appsettings.json, do we want it to be taken into account even if `OC.Localization` is not enabled? If no that's okay, if yes the config binding should not be kept in the `OC.Localization` startup.

hishamco commented:
Agree with you @jtkech as I said before, but using the name `UseSystemCultureSettings` or `UseSystemSettings` with default value `true` will be much better if all agree

> do we want it to be taken into account even if OC.Localization is not enabled? If no that's okay, if yes the config binding should not be kept in the OC.Localization startup.

Agree, but let me do some tests and update the PR

jtkech commented on src/OrchardCore/OrchardCore/Modules/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -126,6 +126,15 @@ private static void AddDefaultServices(OrchardCoreBuilder builder)
+                        .Get<CultureOptions>();
+
+                    return cultureOptions;
+                });
```
`builder.ConfigureServices()` is like a module startup in which we can inject services that are already registered as `ShellSettings` and `IShellConfiguration`, this even if we are building a tenant container.

In case of `builder.ConfigureServices()` you can pass to the delegate an `IServiceProvider` parameter that you can use to resolve `IShellConfiguration` and then use one of the regular `s.Configure<CultureOptions>()`.

            builder.ConfigureServices((s, serviceProvider) =>
            {
                s.AddSingleton<LocalLock>();
                s.AddSingleton<ILocalLock>(sp => sp.GetRequiredService<LocalLock>());
                s.AddSingleton<IDistributedLock>(sp => sp.GetRequiredService<LocalLock>());

                var configuration = serviceProvider.GetRequiredService<IShellConfiguration>();
                s.Configure<CultureOptions>(configuration.GetSection("OrchardCore_Localization_CultureOptions"));
            });

Note: I could have also named it `sp` in place of `serviceProvider`, that's would be okay as other `sp` are used in their own closure (if we click go to definition we can see that it makes the distinction), so I used `serviceProvider` just to make a naming distinction, but you can also use `sp`, as you want.

hishamco replied:
I'm not sure why I didn't call `Configuure()` instead?!! May be I was sleepy ;)

hishamco replied:
Seems I didn't notice the second overload

jtkech reviewed (commented):
Just one suggested change, otherwise LGTM

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Localization/CultureScope.cs:
```diff
@@ -3,34 +3,69 @@
 namespace OrchardCore.Localization
 {
+    /// <summary>
+    /// 
```
Seems I need to complete the docs

jtkech commented:
@hishamco Did you tried it? If so I will approve ;)

hishamco commented:
I just asked you in case if you have anything else, I will add a missing docs then do a final test after latest commit

hishamco commented:
@jtkech seems we forgot to apply the same settings in `SetDefaultCulture()`, one more thing that I was did it right but you confused me ;) may be because the naming

`UseSystemSettings` != `useUserOverride`
`IgnoreSystemSettings` == `useUserOverride`

For that I remove the negation as it was and everything is fine

![CultureInfo](https://user-images.githubusercontent.com/3237266/193290987-520bc401-9dfd-42dc-9882-406674a76286.png)

hishamco commented:
I'm think to do a last tweak for the added extension by adding `OrchardCoreRequestLocalizationOptions` for doing the heavy lifting, then we can use as the following:

```csharp
var requestLocalizationOptions = serviceProvider.GetService<IOptions<OrchardCoreRequestLocalizationOptions>>().Value;

requestLocalizationOptions
    .SetDefaultCulture(defaultCulture)
    .AddSupportedCultures(supportedCultures)
    .AddSupportedUICultures(supportedCultures);

app.UseRequestLocalization(requestLocalizationOptions);
```

jtkech commented:
@hishamco 

> @jtkech seems we forgot to apply the same settings in SetDefaultCulture()

Yes as I mentioned here https://github.com/OrchardCMS/OrchardCore/issues/11228#issuecomment-1105795875

> one more thing that I was did it right but you confused me ;) may be because the naming

Just saw that you now use `new CultureInfo(culture, ignoreSystemSettings)` which is wrong as by default when just using `new CultureInfo(culture)` it doesn't ignore system settings, so if it was not working it would mean that something else is wrong. But never sure at 100%, will look at it tonight.

jtkech commented:
Yes I know https://github.com/dotnet/runtime/blob/5481c4bfe6ced9000c05692163a019739c912677/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs#L158-L162

> Because the useUserOverride is true, in our case if ignoreSystemSettings is true it should ignore the default settings

Because the `useUserOverride` is true by default, meaning it will use system settings, and because `ignoreSystemSettings` is false by default (if not configured/overridden), we should use `new CultureInfo(culture, useUserOverride: !ignoreSystemSettings)`, so that it works as before if the option value is not configured (e.g. in appsettings) and not overridden by code (e.g. in a s.Configure<>()).

So if `IgnoreSystemSettings` is configured/overridden to true (which is not the default) the result will be `new CultureInfo(culture, useUserOverride: false)` which is right as here we want to ignore system settings, so we don't want to use user override (meaning we don't want to use system settings).

hishamco commented:
Believe it or not the `IgnoreSystemSettings is confusing especially we you read the `CultureInfo(string bool)` docs ;)

hishamco commented:
I need to do a final quick test again

jtkech commented:
> Believe it or not the IgnoreSystemSettings is confusing especially we you read the CultureInfo(string, bool) docs ;)

Yes I agree ;)

hishamco commented:
Everythings looks as expected

Piedone commented on src/OrchardCore/OrchardCore.Abstractions/Localization/CultureOptions.cs:
```diff
@@ -0,0 +1,12 @@
+public class CultureOptions
+{
+    /// <summary>
+    /// Gets or sets whether to ignore the system culture settings or not.
```
Link to the issue in the summary for context, otherwise it won't be apparent what we're talking about and why we have the option.

Same in `CultureScope`.

hishamco replied:
We didn't add that frequently, may be we could add extra explanation in remarks

Piedone replied:
That's also good, but an issue link lets people see the full context.

Piedone commented on src/OrchardCore/OrchardCore.Localization.Abstractions/OrchardCoreRequestLocalizationOptions.cs:
```diff
@@ -0,0 +1,68 @@
+    /// Creates a new <see cref="OrchardCoreRequestLocalizationOptions"/> with adefault values.
+    /// </summary>
+    public OrchardCoreRequestLocalizationOptions() : this(ignoreSystemSettings: false)
+    {
+
+    }
```
```suggestion
    {
    }
```

Piedone replied:
Did you apply this change? Can't see it.

Piedone commented on src/OrchardCore/OrchardCore.Localization.Abstractions/OrchardCoreRequestLocalizationOptions.cs:
```diff
@@ -0,0 +1,68 @@
+    private readonly bool _useUserOverride;
+
+    /// <summary>
+    /// Creates a new <see cref="OrchardCoreRequestLocalizationOptions"/> with adefault values.
```
```suggestion
    /// Creates a new <see cref="OrchardCoreRequestLocalizationOptions"/> with default values.
```

hishamco replied:
Nah, typo ;)

Piedone commented on src/OrchardCore/OrchardCore.Localization.Abstractions/OrchardCoreRequestLocalizationOptions.cs:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    /// <summary>
+    /// Creates a new <see cref="OrchardCoreRequestLocalizationOptions"/> with adefault values and ability to ignore system settings.
```
```suggestion
    /// Creates a new <see cref="OrchardCoreRequestLocalizationOptions"/> with default values and ability to ignore system settings.
```

Piedone reviewed (changes requested):
Nice work!

Piedone commented on src/OrchardCore/OrchardCore.Abstractions/Localization/CultureScope.cs:
```diff
@@ -6,6 +6,7 @@ namespace OrchardCore.Localization
     /// <summary>
     /// Represents a scope that you can change the current culture within.
     /// </summary>
+    /// <remarks>The scope disallow the current culture depends on local computer settings by default. For more information refer to https://github.com/OrchardCMS/OrchardCore/issues/11228</remarks>
```
This is a really long line, add a line break in between. Same for CultureOptions.

hishamco commented:
FYI I created a PR in aspnet repo while I'm working on this feature, finally it got merge. I might need to open a draft PR to change the API with the one that will be built-in in .NET 8.0

https://github.com/dotnet/aspnetcore/pull/44282

Piedone commented:
Great!

alafi commented:
I believe this PR introduced this issue #12610

hishamco commented:
Thanks @alafi I will check the related issue
